### PR TITLE
Embed ActorID inside task and object IDs

### PR DESCRIPTION
**Pull request by stephanie-wang**
## What do these changes do?

Shortens ActorID to 4 bytes (up for debate) and embeds this ID into the IDs of tasks destined for that actor and objects created by those tasks. Tasks/objects that are not associated with an actor will contain `0xff` for those 4 bytes.

This PR also refactors `id.h` a bit to hopefully make it a little safer to add new `UniqueIDs` in the future that may have different sizes. Specifically, it adds a runtime check whenever a `UniqueID` is created from a binary string that the string has the right length, requires that the `UniqueID` size is specified as part of the type, and adds static checks that all types have the right size.

After this, we will be able to determine what actor a task/object is destined for without having to retrieve the task spec.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: src/ray/common/constants.h [hunk 1]
```diff
@@ -13,8 +13,9 @@ constexpr int kObjectIdIndexSize = 32;
 static_assert(kObjectIdIndexSize % CHAR_BIT == 0,
               "ObjectID prefix not a multiple of bytes");
 
+constexpr int64_t kActorIdSize = 4;
```
Comment (zhijunfu) [inline src/ray/common/constants.h]: Java currently uses the return object ID of the actor creation task as actor ID, that would need to be changed as well? Similar changes are needed for `TaskInterface` in C++ core worker which does translations between ObjectID and ActorID.

**Comment by jovany-wang** [comment]
I'm worrying about it's not enough safe to generate a 4 bytes ID randomly.

**Comment by raulchen** [comment]
I'm not sure how useful this would be.
In which cases do you think we'll want to quickly get the actor id from a task/object id?

Besides task id, I think we may also want to embed job id and potentially batch id (for batch GC) in object id. So I think we'll need to carefully choose which are mostly important to avoid object id growing too big.

**Comment by stephanie-wang** [comment]
> I'm not sure how useful this would be.
> In which cases do you think we'll want to quickly get the actor id from a task/object id?

Sorry, I should've written this in the PR description, but the main reason for this change is so we can remove task leases for actor tasks. In the future, it could also be used for other optimizations (such as retrieving the return value of an actor task quickly).

> Besides task id, I think we may also want to embed job id and potentially batch id (for batch GC) in object id. So I think we'll need to carefully choose which are mostly important to avoid object id growing too big.

That's true, I'm also worried about this. Let me think about whether we can implement the change to remove task leases for actor tasks without this PR.

**Comment by stephanie-wang** [comment]
Closing for now while we decide the best course of action for removing task leases for actors.

**Comment by pcmoritz** [comment]
If not having the actor ID part of the task/object IDs makes it much harder
to remove task leases, we can consider a different approach to fix the
collisions, e.g. generate unique int32 actor IDs with some kind of counting
scheme.

**Comment by stephanie-wang** [comment]
I think in terms of code complexity, it's not too much worse, but it will definitely be more expensive. It will require at least one GCS lookup for every task that might have been lost so that we can determine its actor ID. And we'd have to do this even during normal operation when there are no node failures because evicted objects also count as "lost".

**Pull request closed without merging**